x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/analysis/PostAnalyzer.java

```diff
@@ -49,10 +48,16 @@ public LogicalPlan postAnalyze(LogicalPlan plan, EqlConfiguration configuration)
             Holder<Boolean> hasJoin = new Holder<>(Boolean.FALSE);
 
             Source projectCtx = synthetic("<implicit-project>");
-            if (plan.anyMatch(Sequence.class::isInstance)) {
-                hasJoin.set(Boolean.TRUE);
+            if (plan.anyMatch(x -> x instanceof Sample)) {
+                plan = plan.transformUp(KeyedFilter.class, k -> {
+                    hasJoin.set(Boolean.TRUE);
+                    Project p = new Project(projectCtx, k.child(), k.keys());
+                    return new KeyedFilter(k.source(), p, k.keys(), k.timestamp(), k.tiebreaker());
+                });
+            } else {
```

Comment:
The new piece of code has different semantics then the previous one - the projection is now applied to all keyed filter not just those under Sequence (including Sample) which is too broad.
Second there's the issue of code duplication since both branches create new objects and do the transformUp - I suggest making them more compact:

```
 if (plan.anyMatch(Sequence.class::isInstance)) {
    hasJoin.set(Boolean.TRUE);
    final boolean isSample = Sequence.class.isInstance(plan);
    plan = plan.transformUp(KeyedFilter.class, k -> {
          var keys = isSample ? k.keys() : k.extractionAttributes();
          Project p = new Project(projectCtx, k.child(), keys);
          ...
    }
```

Essentially the if filter down into the plan transformUp method as oppose to where it is right now.